Improved color dimmer support

emontnemery opened the pull request:
As discussed in #2078, improved support for color LED dimmers.

2018-03-29:
Updated binaries matching commit fae2ff0c78c0879b2ac61afe439ea8910df0ac40:
- Linux x86 64 binary [domoticz_linux_x86_64.tgz](http://montnemery.com/domoticz_rgbww/linux/133.1/fae2ff0c78c0879b2ac61afe439ea8910df0ac40/domoticz_linux_x86_64.tgz)
- Windows binary: [domoticz_windows_x86.zip](http://montnemery.com/domoticz_rgbww/windows/1.0.139/fae2ff0c78c0879b2ac61afe439ea8910df0ac40/domoticz_windows_x86.zip)

**Summary:**
- Color picker for dimmer updated with adjustable color temperature
- Possible to set colors other than fully saturated from Web UI
- Set color of device is stored in database and available through JSON API
- Full color data used internally by Domoticz instead of hue + white level
- Hue stored in existing scenes and timers is translated during database upgrade
- New Subtypes added to Limitlessled:
  - WW (white with adjustable color temperature)
  - RGBWZ - RGB + White, where RGB and White can be turned on at the same time
  - RGBWWZ - RGB + Warm white + Cold White, where RGB and White can be turned on at the same time

**API changes:**
- Field "Color" added to "devices" web API
- Filed "Color" added to MQTT status update from Domoticz
- Web API command "setcolbrightnessvalue" now accepts JSON formatted color data as well as hue
- Web API commands to set/get timers and set/get scene devices now accepts JSON formatted color data instead of hue
- JSON formatted color data passed to Python plugins instead of Hue

**Open issues / Not yet implemented:**
- Consider improving usability of the updated color picker in custom mode, it is not very intuitive, maybe grouping of the sliders would improve it
- Testing of the various HW is needed!

**HW status:**
Arilux:
- RGBWW / CW_WW hardware: Yes, according to comments in the source code
- RGBWW / CW_WW Domoticz support: No
- Test results: Not tested

Blebox:
- RGBWW / CW_WW hardware: No RGBWW devices according to homepage: http://blebox.eu/en/12-products
- RGBWW / CW_WW Domoticz support: No
- Test results: Not tested

EHouse:
- RGBWW / CW_WW hardware: ?
- RGBWW / CW_WW Domoticz support: No
- Test results: Not tested

Limitless:
- RGBWW / CW_WW hardware: Yes, RGBWW lights
- RGBWW / CW_WW Domoticz support: Partial, RGBWW and adjustable color temperature type light can be added from Web UI.
- Test results: Tested OK by @gijsje and @poudenes

MySensors:
- RGBWW / CW_WW hardware: ?
- RGBWW / CW_WW Domoticz support: No
- Test results: Not tested

PhilipsHue:
- RGBWW / CW_WW hardware: Yes
- RGBWW / CW_WW Domoticz support: Yes, support for detection of both RGBWW and CW_WW type lights added
- Test results: Tested OK by myself (groups not tested)

RFLink:
- RGBWW / CW_WW hardware: ?
- RGBWW / CW_WW Domoticz support: No
- Test results: Changing color works, brightness does not (tested by @poudenes)

XiaomiGateway:
- RGBWW / CW_WW hardware: ?
- RGBWW / CW_WW Domoticz support: No
- Test results: Not tested

Yeelight:
- RGBWW / CW_WW hardware: Yes
- RGBWW / CW_WW Domoticz support: Partial support for detection of RGBWW and WW type bulbs added, but web UI needs to be updated. 
- Support for writing color data to Domoticz database needs to be added.
- Web UI needs to be updated
- Test results: Tested OK by @adrianmihalko

ZWave:
- RGBWW / adjustable color temperature dimmers: Yes, 8 channel color dimmer supported by protocol
- RGBWW / CW_WW Domoticz support: No
- Test results: Tested OK by @mdagoreau
- Workaround for adjustable color temperature added for Zipato Bulb 2
- Outstanding issue is to improve ZWave color dimmer discovery:
Supported color channels is reported in response to command SWITCH_COLOR_SUPPORTED_GET, and this command is automatically sent by OpenZWave and the result is stored as a member variable of OpenZWave::Color class and is be available for the user as well.
This data should be used to add discovered devices as WW / RGB / RGBWZ / RGBWWZ type device.
Since I don't have any ZWave HW, I can't implement such a big change myself.

pTypeLighting5.sTypeTRC02:
- According to some code in mainworker.cpp, this HW supports color
Does anyone have this device, and is it currently working?
- Test results: Not tested

emontnemery committed:
Improved color dimmer support

poudenes commented:
@emontnemery everything work using V6 box. Rgbww and rgbw. Color picker with brightness and Kelvin with brightness work. I can select "set white" and use brightness.  Hopefully rflink will go support rgbww as well so I can control it with rflink without V6 box. (Different discussion)

emontnemery commented:
@gizmocuz Please review, I think this is ready for merge now. There are still some debug prints to aid debugging, I plan to remove them some time after merge once reported regressions / bugs have been sorted out. If you want, I can tag these prints with `//TODO: Temporary RGBWW debug print, remove` or something so they're not forgotten. There was a lot of help with testing and debugging in #2078, remaining issues should be limited to the HW types which has not been tested.

@dnpwwo Plugins now get a serialized JSON color object. This is not backwards compatible, but I can't find any plugin actually implementing support for color dimmer except for this one which seems not actively developed: https://github.com/zaraki673/MiPow-PlayBulb/ (which @999LV is working on a replacement for) so maybe that's OK?

poudenes commented:
@emontnemery will you proceed improvements for White + Brightness?

emontnemery commented:
@poudenes Do you mean the issue with RFLink? Everything is working as expected with the V6 box, right?

poudenes commented:
Color Picker  + Brightness Works
Kelvin Picker + Brightness Works
Kelvin Selectors (0,25,50,75,100) works
setWhite works, but you cannot change brightness for only white

In Beta V3.9052 there is a White Slider. Thought that will be the brightness slider for only white.

emontnemery commented:
@poudenes For Milight V6 RGBWW, "Set White" button is the same as this setting in kelvin picker:
![image](https://user-images.githubusercontent.com/14281572/37915316-8621d24c-3119-11e8-8f4c-62e55c7622cf.png)
I will remove the "set white" button and kelvin percentage buttons, they're just confusing.

poudenes commented:
@emontnemery confirmed. Test it and yes lowest part of Kelvin is white and work with Brightness (WORK)

emontnemery commented:
@poudenes Thanks! I removed the "Set White" button, I think it's only confusing.

gizmocuz commented:
Great patch !! Thanks all for the hard work !

Unfortunately i would like to test as well (specifically the zwave part), but time is my problem.

Also this can not be merged because of this:

Travis and appveyour file is updated to deploy to my machine, please ignore
domoticz.js temporarily added to be able to review changes, please merge as .gz

And the biggest challenge is that i would like to merge another patch that simplifies a lot of html/java code as well, that has changes in domoticz.js (.gz)

https://github.com/domoticz/domoticz/pull/2240

emontnemery commented:
@gizmocuz OK, please go ahead and merge #2240 first if it's easier, then I'll to a rebase + forced push of this branch after the merge.
I'll also revert appveyor settings, travis settings and non-compressed domoticz.js

stas-demydiuk commented:
@gizmocuz what do you think about moving `domoticz.js.gz` into regular `domoticz.js`, in this case it will allow us to have proper history of this file and definitely simplify merging/rebasing and conflict resolving as this file contains a lot of domoticz related logic basically?

adrianmihalko commented:
+1 for regular domoticz.js. On localhost (where Domoticz is usually running at) there is no big advantage for compressed javascript and everything is cached anyway.

emontnemery commented:
Agree about uncompressed domoticz.js.. 
In this PR I actually gzip domoticz.js during Travis / Appveyor build, maybe ok as interim solution?

gizmocuz commented:
@stas-demydiuk, yes, in fact, all compressed files are (very) soon uncompressed.
It makes sense to have them compressed for arm systems with a low cpu, but this could be done internally by either the software based on a hash, or during packaging

gizmocuz commented:
@emontnemery , temperature patch has been merged.
Please feel free to uncompress/add domoticz.js and delete the .gz one
This will also force me in the weekend to look at the compression part ;)

emontnemery commented:
@gizmocuz Great! I've squashed and rebased and will push a new commit reverting the travis and appveyor setting in a moment.
Edit: The latest commit has reverted travis and appveyor settigns to Domoticz standard.

emontnemery committed:
Revert appveyor and travis settings

gizmocuz commented on main/SQLHelper.cpp:
```diff
@@ -2419,6 +2419,92 @@ bool CSQLHelper::OpenDatabase()
 			safe_query("UPDATE Hardware SET Mode1 = 5000 WHERE Type = %d", HTYPE_DenkoviSmartdenIPInOut);
 			safe_query("UPDATE Hardware SET Mode1 = 5000 WHERE Type = %d", HTYPE_DenkoviSmartdenLan);
 		}
+		if (dbversion < 123)
```
You need to use 124 ... i used 123 in a previous patch ;)

gizmocuz commented:
Thanks, did you also removed the domoticz.js.gz file ?
The SQL version number to use for changes is currently 124 (i used 123 previously)

One remark, RGB_WW ... is type is a bit unclear, maybe just rename that to RGB_W ?

(because there are devices that have cold white instead of warm)

emontnemery committed:
Fix merge error in SQLHelper

emontnemery commented:
@gizmocuz Thanks, I fixed the merge mistake in SQLHelper. Yes, www/js/domoticz.js.gz is removed and replaced with www/js/domoticz.js

I'm not sure I understand your comment about RGB_W, the following types are defined in hardware/hardwaretypes.h:
```
#define sTypeLimitlessRGBW		0x01 // RGB + white, either RGB or white can be lit
#define sTypeLimitlessRGB		0x02 // RGB
#define sTypeLimitlessWhite		0x03 // Monochrome white
#define sTypeLimitlessRGBWW		0x04 // RGB + warm white + cold white, either RGB or white can be lit
#define sTypeLimitlessLivCol		0x05
#define sTypeLimitlessRGBWZ		0x06 // Like RGBW, but allows combining RGB and white
#define sTypeLimitlessRGBWWZ		0x07 // Like RGBWW, but allows combining RGB and white
#define sTypeLimitlessWW		0x08 // Warm white + Cold white
```

gizmocuz commented:
Thanks !!

We should rename those 'Limitetless' types soon, like sTypeRGB (remove the Limitless)

It's a bit  confusing (for me)

#define sTypeLimitlessRGBWW		0x04 // RGB + warm white + cold white, either RGB or white can be lit

WW = Warm White
CW = Cold White

Above here it means Warm+Cold, so the type should be RGB_CW_WW ?

HLTYPE_WW, is this also RGB + warm white + cold white ?

Maybe also rename this HL = ? HLTYPE_Warm_White

Also the sType_RGB_Warm_and_Cold_White

I know this sounds large... it is, but when you will look at this code a year later, your happy with it

emontnemery commented:
OK, so how about this:

hardware/hardwaretypes.h:
```
#define pTypeColorSwitch		0xF1
#define sTypeColor_RGB_W		0x01 // RGB + white, either RGB or white can be lit
#define sTypeColor_RGB			0x02 // RGB
#define sTypeColor_White		0x03 // Monochrome white
#define sTypeColor_RGB_CW_WW		0x04 // RGB + cold white + warm white, either RGB or white can be lit
#define sTypeColor_LivCol		0x05
#define sTypeColor_RGB_W_Z		0x06 // Like RGBW, but allows combining RGB and white
#define sTypeColor_RGB_CW_WW_Z		0x07 // Like RGBWW, but allows combining RGB and white
#define sTypeColor_CW_WW		0x08 // Cold white + Warm white

typedef struct _tColorSwitch {
...
} _tColorSwitch;
```
hardware/PhilipsHue.h
```
enum _eHueLightType
{
	HLTYPE_NORMAL,
	HLTYPE_DIM,
	HLTYPE_RGBW,
	HLTYPE_SCENE,
	HLTYPE_CW_WW,
	HLTYPE_RGB_CW_WW
};
```

gizmocuz commented:
Yes! Excellent, very clear !

emontnemery committed:
Rename 'LimitLess' to 'Color'

emontnemery commented:
OK, fixed now.

gizmocuz commented:
Thanks for this great patch ! Needs a few tiny adjustments which i will perform shortly

gizmocuz commented:
https://github.com/domoticz/domoticz/commit/09e5d171b79906d561b02fb9dff8030dd4b1f7a6

Thunderbolt commented:
Some yeelight lamps (ceiling lamp for me) have a night light mode. Could that be implemented?

This mode has a much lower brightness range compared to the normal mode.
Another issue, when the ceiling lamp is in night light mode and circle in WW, changing the brightness, changes the mode, too. Which means, the lamp gets the much higher brightness range.
So at the moment, it's not possible to change the brightness in night light mode. 
But when the circle is in RGB, the brightness change doesn't change the night light mode to normal. Now it is possible to change the brightness in this lower range.

I'm not talking about the button "Night Light" under edit, which sets the lowest brightness in normal mode.

jonferreira commented:
@emontnemery fantastic work, lots of time and effort on this one!

I'm now seeing an odd behavior with Hue and MQTT though.

If I turn On / Off / Change Colour via Domo it all works perfectly fine.
If I do it via Hue App / Hue Remote, Domo UI does update but doesn't seem to notify MQTT - bit ofa bummer as it means a big part of my logic isn't working :(

emontnemery commented:
@Thunderbolt If you set to lowest possible dim level in the UI, is it less bright than night mode?
@jonferreira Did you get MQTT notifications when changing state through Hue App before this was merged?

jonferreira commented:
@emontnemery It did yes, used it all the time

emontnemery commented:
@jonferreira OK. Please create a new issue for this.

jonferreira commented:
done.

thanks @emontnemery

morilm commented:
Hi there. Sorry I haven't been able to test anything before due to amount of work

Tested MagicHome controller (IP based) using Arilux HW (as recomended in the thread https://www.domoticz.com/forum/viewtopic.php?t=11479

Light can be swithed on/off.  Color picker work but changing dim doesn't.  On dashboard brings color to withe while on color picker doesn't change value at all.

Hope it helps

Miguel
